### PR TITLE
[server] use options from last instances on start

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
This change makes the workspace start reusing what has been used for IDE and workspace class on the previous workspace instance.

I think this change can land on its own. In a larger second PR we would allow users to start an existing workspace with options, so users can change/override what IDE and workspace class to use on subsequent workspace starts.
## Related Issue(s)

Fixes [WEB-189](https://linear.app/gitpod/issue/WEB-189)

## How to test
Create a workspace with non default IDE and was class. Stop the workspace and restart it. It should still use the previously selected IDE and class.
#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - se-restart-options</li>
	<li><b>🔗 URL</b> - <a href="https://se-restart-options.preview.gitpod-dev.com/workspaces" target="_blank">se-restart-options.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

File: components/server/src/workspace/workspace-starter.ts [hunk 1]
```diff
@@ -275,9 +275,21 @@ export class WorkspaceStarter {
                 }
             }
 
-            const ideConfig = await this.resolveIDEConfiguration(ctx, workspace, user, options.ideSettings);
+            let ideSettings = options.ideSettings;
+
+            // if no explicit ideSettings are passed, we use the one from the last workspace instance
+            if (lastValidWorkspaceInstance) {
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/server/src/workspace/workspace-starter.ts:281]: this happens already down the line

[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/server/src/workspace/workspace-starter.ts:281]: It is called `getWorkspaceClassForInstance` 
https://github.com/gitpod-io/gitpod/blob/da5dcf73cbc3844dd5d68a7e246dff7e348031df/components/server/src/workspace/workspace-starter.ts#L859-L869)

[the remark this answers is not in the document]
Comment (easyCZ) [reply components/server/src/workspace/workspace-starter.ts:281]: That's just the class tho, there are are options - IDE, etc

[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/server/src/workspace/workspace-starter.ts:281]: validation of passed ide config happens in `resolveIDEConfiguration`

File: components/ide-service/pkg/server/server.go [hunk 2]
```diff
@@ -351,6 +354,23 @@ func (s *IDEServiceServer) ResolveWorkspaceConfig(ctx context.Context, req *api.
 			resp.IdeImageLayers = append(resp.IdeImageLayers, desktopImageLayer)
 			resp.IdeImageLayers = append(resp.IdeImageLayers, desktopUserImageLayers...)
 		}
+
+		// we are retunring the actually used ide name here, which might be different from the user's choice
```
Comment (svenefftinge) [inline components/ide-service/pkg/server/server.go:358]: @mustard-mh Could you take a look? I have extended this service to return the actual ide Settings again, so we can store what was used in the workspace_instance and use it for the next workspace start.

Comment (akosyakov) [reply components/ide-service/pkg/server/server.go:358]: maybe add some tests

File: components/ide-service/pkg/server/server.go [hunk 3]
```diff
@@ -287,6 +287,7 @@ func (s *IDEServiceServer) ResolveWorkspaceConfig(ctx context.Context, req *api.
 
 		userIdeName := ""
 		useLatest := false
+		resultingIdeName := ideSettings.DefaultIde
```
Comment (akosyakov) [inline components/ide-service/pkg/server/server.go:290]: ideSettings can be nil?

**Comment by akosyakov** [comment]
Is it fine to deploy ide-service and server in any order?

File: components/ide-service/pkg/server/server.go (hunk 2 above)
Comment (svenefftinge) [reply components/ide-service/pkg/server/server.go:358]: I need to update the golden files

**Comment by svenefftinge** [comment]
> Is it fine to deploy [ide-service](https://linear.app/gitpod/issue/IDE-service) and server in any order?

I think so. The only use of the newly introduced field in the result type is guarded ([code](https://github.com/gitpod-io/gitpod/pull/17298/files#diff-b1c591e18af598ccba93b570ce7c97dece0988c9109a6d3be82bbe0df1fbd619R832)).

**Comment by akosyakov** [comment]
Sounds good. I would like to try, but build fails. :(

**Comment by svenefftinge** [comment]
@akosyakov the preview env now works. I verified that restarting a workspace opens with the same IDE. Can you take another look?

File: components/server/src/workspace/workspace-starter.ts [hunk 4]
```diff
@@ -818,6 +829,15 @@ export class WorkspaceStarter {
                     useLatest: user.additionalData?.ideSettings?.useLatestVersion,
                 },
             };
+            if (ideConfig.ideSettings && ideConfig.ideSettings.trim() !== "") {
+                try {
+                    const ideSettings: IDESettings = JSON.parse(ideConfig.ideSettings);
+                    configuration.ideConfig!.desktopIdeAlias = ideSettings.defaultIde;
```
Comment (akosyakov) [inline components/server/src/workspace/workspace-starter.ts:835]: @svenefftinge Should we remove desktopIdeAlias? It is not used anywhere on main.

And instead add a new property like `ide` to reflect that it is any IDE.

**Review by akosyakov — APPROVED** [review]
it works, but i think we should use another property to avoid confusion: https://github.com/gitpod-io/gitpod/pull/17298/files#r1173888796

/hold

**Comment by svenefftinge** [comment]
Thanks Anton! I have updated the shape as suggested.

/unhold

**Pull request merged**